### PR TITLE
CLDR-15034 kbd: Keyboard 3.0 DTD and Data Files

### PR DESCRIPTION
**Pull request by srl295**
*Note*: the spec is already landed in #3111

- remove old Keyboard DTD and data files
- add new Keyboard DTD and Data Files
- add Keyboard tools
- originally many commits and PRs

Co-authored-by: Marc Durdin <marc@durdin.net>
(cherry picked from commit 1716d10929d39b19cc8f03f6feba8d4f789bacd3)

CLDR-15034

- [ ] This PR completes the ticket.

An earlier version of this PR included the flatten tool, this was split out to #3117

ALLOW_MANY_COMMITS=true

-----

### Review note
- one of the commits is literally just deleting the old DTD and data files
- Feedback welcome, but I'd like to address any keyboard related feedback in _subsequent PRs_.
- Please look especially at the .java files changed here to make sure there aren't any issues.

### CODE REVIEW

**Comment by srl295** [comment]
may depend on #3108

**Comment by miloush** [comment]
We mentioned in the preview/PRI that the element/attribute names are subject to change for the final version. Do we want to go through them before this is merged into main?

**Comment by srl295** [comment]
> We mentioned in the preview/PRI that the element/attribute names are subject to change for the final version. Do we want to go through them before this is merged into main?

They can still be revised. We could revise them first, but it makes some sense to merge what matches the PRI/preview and then make a separate pr to revise a name. 

I have a separate issue to remove the techpreview label.

**Comment by srl295** [comment]
removed the schema check

```yaml
      - name: Keyboard Schema Check
        run: >
          mvn -s .github/workflows/mvn-settings.xml --file tools/pom.xml -pl cldr-code
          exec:java -Dexec.mainClass=org.unicode.cldr.util.XMLValidator
          -Dexec.cleanupDaemonThreads=false
          -Dexec.args="--schema=keyboards/dtd/ldmlKeyboard.xsd keyboards/3.0"
          && echo 'No XML BOM detected'
      - name: Keyboard Test Check
        run: >
          mvn -s .github/workflows/mvn-settings.xml --file tools/pom.xml -pl cldr-code
          exec:java -Dexec.mainClass=org.unicode.cldr.util.XMLValidator
          -Dexec.cleanupDaemonThreads=false
          -Dexec.args="--schema=keyboards/dtd/ldmlKeyboardTest.xsd keyboards/test"
          && echo 'No XML BOM detected'
```

**Comment by srl295** [comment]
- for now I backed out the semver and bcp47 pending #3109 landing. I'll bring those back once that lands.

#3110 has already landed.

**Comment by srl295** [comment]
Builds now, ready to review and merge.

**Comment by srl295** [comment]
reinstated @MATCH DTD now that #3109  landed.

File: .github/workflows/maven.yml [hunk 1]
```diff
@@ -126,6 +126,20 @@ jobs:
...
+          -Dexec.cleanupDaemonThreads=false
+          -Dexec.args="--schema=keyboards/dtd/ldmlKeyboard.xsd keyboards/3.0"
+          && echo 'Schema failed, rerun GenerateDtd'
+      - name: Keyboard Test Schema Check
+        run: >
+          mvn -s .github/workflows/mvn-settings.xml --file tools/pom.xml -pl cldr-code
+          exec:java -Dexec.mainClass=org.unicode.cldr.util.XMLValidator
+          -Dexec.cleanupDaemonThreads=false
+          -Dexec.args="--schema=keyboards/dtd/ldmlKeyboardTest.xsd keyboards/test"
+          && echo 'Schema failed, rerun GenerateDtd'
```
Comment (srl295) [inline .github/workflows/maven.yml:142]: Two new tests against the .xsd

File: tools/cldr-code/src/main/java/org/unicode/cldr/tool/ChartDtdDelta.java [hunk 2]
```diff
@@ -220,7 +220,6 @@ public void writeContents(FormattedFileWriter pw) throws IOException {
     static {
         FIRST_VERSION.put(DtdType.ldmlBCP47, "1.7.2");
         FIRST_VERSION.put(DtdType.keyboard, "22.1");
-        FIRST_VERSION.put(DtdType.platform, "22.1");
```
Comment (srl295) [inline tools/cldr-code/src/main/java/org/unicode/cldr/tool/ChartDtdDelta.java:223]: platform was removed.

File: tools/cldr-code/src/main/java/org/unicode/cldr/tool/ShowDtdDiffs.java [hunk 3]
```diff
@@ -31,7 +31,6 @@ public class ShowDtdDiffs {
     static {
         FIRST_VERSION.put(DtdType.ldmlBCP47, "1.7.2");
         FIRST_VERSION.put(DtdType.keyboard, "22.1");
-        FIRST_VERSION.put(DtdType.platform, "22.1");
```
Comment (srl295) [inline tools/cldr-code/src/main/java/org/unicode/cldr/tool/ShowDtdDiffs.java:34]: platform was removed

File: tools/cldr-code/src/main/java/org/unicode/cldr/util/CLDRPaths.java [hunk 4]
```diff
@@ -191,16 +191,7 @@ public enum DIRECTORIES {
...
-        common_validity(DtdType.supplementalData),
-
-        keyboards_android(DtdType.keyboard, DtdType.platform),
-        keyboards_chromeos(DtdType.keyboard, DtdType.platform),
-        keyboards_dtd(DtdType.keyboard, DtdType.platform),
-        keyboards_osx(DtdType.keyboard, DtdType.platform),
-        keyboards_und(DtdType.keyboard, DtdType.platform),
-        keyboards_windows(DtdType.keyboard, DtdType.platform),
-        ;
-
```
Comment (srl295) [inline tools/cldr-code/src/main/java/org/unicode/cldr/util/CLDRPaths.java:203]: deleted

File: tools/cldr-code/src/main/java/org/unicode/cldr/util/ElementAttributeInfo.java [hunk 5]
```diff
@@ -75,22 +76,29 @@ public static final ElementAttributeInfo getInstance(String commonDirectory, Dtd
                             DtdType.keyboard,
                             new ElementAttributeInfo(
                                     canonicalCommonDirectory
-                                            + "/../keyboards/android/ar-t-k0-android.xml",
+                                            + "/../keyboards/3.0/fr-t-k0-azerty.xml",
```
Comment (srl295) [inline tools/cldr-code/src/main/java/org/unicode/cldr/util/ElementAttributeInfo.java:79]: different example ile

File: tools/cldr-code/src/main/java/org/unicode/cldr/util/ElementAttributeInfo.java [hunk 6]
```diff
@@ -75,22 +76,29 @@ public static final ElementAttributeInfo getInstance(String commonDirectory, Dtd
...
             }
         }
-        return result.get(dtdType);
+        final ElementAttributeInfo eai = result.get(dtdType);
+        if (eai == null) {
+            throw new NullPointerException(
+                    "ElementAttributeInfo.getInstance(…,"
+                            + dtdType.name()
+                            + ") returns null, please update this function");
+        }
```
Comment (srl295) [inline tools/cldr-code/src/main/java/org/unicode/cldr/util/ElementAttributeInfo.java:100]: an attempt to be more helpful

File: tools/cldr-code/src/test/java/org/unicode/cldr/unittest/TestBasic.java [hunk 7]
```diff
@@ -1400,11 +1404,8 @@ public void TestDtdCompatibility() {
                     switch (type) {
                         case ldmlBCP47:
                         case ldmlICU:
-                            tooOld = version.isOlderThan(CldrVersion.v1_7_2);
-                            break;
                         case keyboard:
-                        case platform:
-                            tooOld = version.isOlderThan(CldrVersion.v22_1);
+                            tooOld = version.isOlderThan(CldrVersion.from(type.firstVersion));
```
Comment (srl295) [inline tools/cldr-code/src/test/java/org/unicode/cldr/unittest/TestBasic.java:1408]: make this more robust

File: tools/cldr-code/src/test/java/org/unicode/cldr/unittest/TestDtdData.java [hunk 8]
```diff
@@ -304,11 +306,13 @@ private void checkEmpty(
...
-                                        + "\t|| "
-                                        + showPath(parents)
-                                        + "||path has neither value NOR value attributes||");
+                        if (!ALLOWED_EMPTY_NO_VALUE_PATHS.contains(showPath(parents))) {
+                            errln(
+                                    type
+                                            + "\t|| "
+                                            + showPath(parents)
+                                            + "||DTD has neither value NOR value attributes (only special or deprecated children)||");
+                        }
```
Comment (srl295) [inline tools/cldr-code/src/test/java/org/unicode/cldr/unittest/TestDtdData.java:315]: This and following are improvements, made as a result of many challenges running these tests, to the output text

File: tools/cldr-code/src/test/java/org/unicode/cldr/unittest/TestDtdData.java [hunk 9]
```diff
@@ -433,6 +451,11 @@ public void TestNewDtdData() {
     //
     //    }
 
+    /** paths that can be empty elements. Each item starts with '!' because of showPath. */
+    static final Set<String> ALLOWED_EMPTY_NO_VALUE_PATHS =
+            Collections.unmodifiableSet(
+                    new HashSet<>(Arrays.asList("!//keyboardTest/tests/test/backspace")));
+
```
Comment (srl295) [inline tools/cldr-code/src/test/java/org/unicode/cldr/unittest/TestDtdData.java:458]: The DTD already marks it as empty. This just keeps the test from complaining.

File: tools/cldr-code/src/test/java/org/unicode/cldr/unittest/TestLocale.java [hunk 10]
```diff
@@ -192,7 +192,11 @@ public void TestLocalePartsValidity() {
             if (parent.contains("transform")
                     || parent.contains("bcp47")
                     || parent.contains("supplemental")
-                    || parent.contains("validity")) {
+                    || parent.contains("validity")
+                    || parent.contains(
+                            "keyboards/import") // exclude keyboard imports: not locale IDs
+                    || parent.contains(
+                            "keyboards/test")) { // exclude keyboard test files: not locale IDs
```
Comment (srl295) [inline tools/cldr-code/src/test/java/org/unicode/cldr/unittest/TestLocale.java:199]: these files do not have locale IDs for names

File: tools/cldr-code/src/test/java/org/unicode/cldr/unittest/TestPaths.java [hunk 11]
```diff
@@ -379,10 +379,8 @@ public void show(int inclusion) {
             for (DtdType dtdType : DtdType.values()) {
                 if (dtdType == DtdType.ldmlICU
                         || (inclusion <= 5
-                                && dtdType
-                                        == DtdType
-                                                .platform)) { // keyboards/*/_platform.xml won't be
-                    // in the list for non-exhaustive runs
+                                && false)) { // keyboards/*/_platform.xml won't be in the list for
+                    // non-exhaustive runs
```
Comment (srl295) [inline tools/cldr-code/src/test/java/org/unicode/cldr/unittest/TestPaths.java:383]: - [ ] OK, this is a TODO, should skip both keyboard types

File: tools/cldr-code/src/test/java/org/unicode/cldr/unittest/TestPaths.java [hunk 12]
```diff
@@ -454,7 +452,10 @@ public void TestNonLdml() {
             String dirPath = CLDRPaths.BASE_DIRECTORY + directory;
             for (String fileName : new File(dirPath).list()) {
                 File dir2 = new File(dirPath + fileName);
-                if (!dir2.isDirectory() || fileName.equals("properties") // TODO as flat files
+                if (!dir2.isDirectory()
+                        || (dir2.getName().equals("import")
+                                && directory.equals("keyboards/")) // has a different root element
```
Comment (srl295) [inline tools/cldr-code/src/test/java/org/unicode/cldr/unittest/TestPaths.java:457]: the 'imports' don't start with a root element of `<keyboard>`, see spec

File: tools/cldr-code/src/test/java/org/unicode/cldr/unittest/TestPaths.java [hunk 13]
```diff
@@ -558,7 +566,7 @@ public void TestNonLdml() {
                                     logln("Non-node values: " + nonFinalValues + "\t" + path);
                                 }
                             }
-                            if (isVerbose()) {
+                            if (true || isVerbose()) {
```
Comment (srl295) [inline tools/cldr-code/src/test/java/org/unicode/cldr/unittest/TestPaths.java:569]: - [ ] TODO debugging, to remove

**Comment by srl295** [comment]
I've addressed all of my own TODOs.

**Comment by srl295** [comment]
Ping… I know there's a lot here, any comments?

**Review by mcdurdin — APPROVED** [review]
Conceptually looks good to me, not seeing any glaring issues. But I am not yet familiar enough with CLDR java code to be very helpful in reviewing that 😄

**Review by btangmu — APPROVED** [review]
I just scanned the java quickly, didn't spot any problems

**Review by macchiati — APPROVED** [review]
Might have to refine later, but ok for now.

**Pull request merged**